wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wdl/WdlWriter.scala

```diff
@@ -0,0 +1,450 @@
+    }
+  }
+
+  implicit val stringPieceWriter: WdlWriter[StringPiece] = new WdlWriter[StringPiece] {
+    override def toWdl(a: StringPiece): String = a match {
+      case a: StringLiteral     => a.value
+      case a: StringPlaceholder => "~{" + a.expr.toWdl + "}"
+    }
+  }
+
+  // Recursive references must be explicit
+  implicit val expressionElementWriter: WdlWriter[ExpressionElement] = new WdlWriter[ExpressionElement] {
+    override def toWdl(a: ExpressionElement) = a match {
+      case a: PrimitiveLiteralExpressionElement => a.toWdl
+      case a: StringExpression => "\"" + a.pieces.map(_.toWdl).mkString + "\""
+      case a: StringLiteral => "\"" + a.value + "\""
+      case a: ObjectLiteral =>
+        "object { " + a.elements.map { pair =>
+          pair._1 + ": " + expressionElementWriter.toWdl(pair._2)
+        }.mkString(", ") + " }"
+      case a: ArrayLiteral =>
+        "[" + a.elements.map(expressionElementWriter.toWdl).mkString(", ") + "]"
+      case a: MapLiteral =>
+        "{ " + a.elements.map { pair =>
+          expressionElementWriter.toWdl(pair._1) + ": " + expressionElementWriter.toWdl(pair._2)
+        }.mkString(", ") + " }"
+      case a: PairLiteral =>
+        s"(${expressionElementWriter.toWdl(a.left)}, ${expressionElementWriter.toWdl(a.right)})"
+      case a: UnaryOperation => a.toWdl
+      case a: BinaryOperation => a.toWdl
+      case a: TernaryIf =>
+        s"if ${expressionElementWriter.toWdl(a.condition)} then ${expressionElementWriter.toWdl(a.ifTrue)} else ${expressionElementWriter.toWdl(a.ifFalse)}"
+      case a: FunctionCallElement => a.toWdl
+      case a: IdentifierLookup => a.identifier
+      case a: IdentifierMemberAccess => a.toWdl
+      case a: ExpressionMemberAccess => s"${expressionElementWriter.toWdl(a.expression)}.${a.memberAccessTail.toList.mkString(".")}"
+      case a: IndexAccess => s"${expressionElementWriter.toWdl(a.expressionElement)}[${expressionElementWriter.toWdl(a.index)}]"
+    }
+  }
+
+  implicit val unaryOperationWriter: WdlWriter[UnaryOperation] = new WdlWriter[UnaryOperation] {
+    override def toWdl(a: UnaryOperation): String = a match {
+      case a: LogicalNot    => s"!(${a.argument.toWdl})"
+      case a: UnaryNegation => "-" + a.argument.toWdl
+      case a: UnaryPlus     => "+" + a.argument.toWdl
+    }
+  }
+
+  implicit val identifierMemberAccessWriter: WdlWriter[IdentifierMemberAccess] = new WdlWriter[IdentifierMemberAccess] {
+    override def toWdl(a: IdentifierMemberAccess): String = {
+      s"${a.first}.${a.second}" + (if (a.memberAccessTail.nonEmpty) {
+        "." + a.memberAccessTail.mkString(".")
+      } else {
+        ""
+      })
+    }
+  }
+
+  implicit val binaryOperationWriter: WdlWriter[BinaryOperation] = new WdlWriter[BinaryOperation] {
+    override def toWdl(a: BinaryOperation) = {
+      def sandwich(opString: String) = s"${a.left.toWdl} $opString ${a.right.toWdl}"
```

Comment:
I wonder whether it's necessary to add parens here to avoid accidentally undoing order of precedence?
eg does `(1+2)*3` get converted to `1+2*3`?

Reply:
Quite possibly. On the other hand, we end up with beauties like
```
Int j = (((a + b) + c) + x)
```
so I have been wondering if there is a smart way to insert parens only as needed